### PR TITLE
Coinswap Upgrade based on Cosmos-SDK v0.46.1

### PR DESCRIPTION
**Pull request by taramakage**
- [x] pass compilation
- [x] pass unit test
- [x] pass simulation
- [ ] pass localnet test

### CODE REVIEW

**Review by aofengli — APPROVED** [review]

File: modules/coinswap/keeper/keeper.go [hunk 1]
```diff
@@ -103,13 +104,22 @@ func (k Keeper) Swap(ctx sdk.Context, msg *types.MsgSwapOrder) error {
...
+	var msg *types.MsgAddLiquidity
+
+	switch i := imsg.(type) {
+	case *types.MsgAddUnilateralLiquidity:
+		return k.addUnilateralLiquidity(ctx, i)
+	case *types.MsgAddLiquidity:
+		msg = i
+	}
+
 	standardDenom := k.GetStandardDenom(ctx)
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:117]: why not add function `AddUnilateralLiquidity`

File: modules/coinswap/keeper/keeper.go [hunk 2]
```diff
@@ -197,8 +207,94 @@ func (k Keeper) addLiquidity(ctx sdk.Context,
...
+
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	poolAddr, err := sdk.AccAddressFromBech32(pool.EscrowAddress)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:226]: no need to check

File: modules/coinswap/keeper/keeper.go [hunk 3]
```diff
@@ -197,8 +207,94 @@ func (k Keeper) addLiquidity(ctx sdk.Context,
...
+	poolId := types.GetPoolId(msg.CounterpartyDenom)
+	pool, exist := k.GetPool(ctx, poolId)
+	if !exist {
+		return sdk.Coin{}, sdkerrors.Wrap(types.ErrReservePoolNotExists, fmt.Sprintf("liquidity pool: %s ", poolId))
+	}
+
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:221]: move to the front of line 211

File: modules/coinswap/keeper/keeper.go [hunk 4]
```diff
@@ -197,8 +207,94 @@ func (k Keeper) addLiquidity(ctx sdk.Context,
...
+func (k Keeper) RemoveLiquidity(ctx sdk.Context, imsg interface{}) (sdk.Coins, error) {
+	var msg *types.MsgRemoveLiquidity
+
+	switch i := imsg.(type) {
+	case *types.MsgRemoveUnilateralLiquidity:
+		return k.removeUnilateralLiquidity(ctx, i)
+	case *types.MsgRemoveLiquidity:
+		msg = i
+	}
+
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:297]: ditto

File: modules/coinswap/keeper/keeper.go [hunk 5]
```diff
@@ -282,6 +378,113 @@ func (k Keeper) removeLiquidity(ctx sdk.Context, poolAddr, sender sdk.AccAddress
...
+	poolId := types.GetPoolId(msg.CounterpartyDenom)
+	pool, exist := k.GetPool(ctx, poolId)
+	if exist != true {
+		return sdk.Coins{}, sdkerrors.Wrap(types.ErrReservePoolNotExists, fmt.Sprintf("liquidity pool: %s ", poolId))
+	}
+
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:393]: ditto

File: modules/coinswap/keeper/keeper.go [hunk 6]
```diff
@@ -282,6 +378,113 @@ func (k Keeper) removeLiquidity(ctx sdk.Context, poolAddr, sender sdk.AccAddress
 	return coins, k.bk.SendCoins(ctx, poolAddr, sender, coins)
 }
 
+func (k Keeper) removeUnilateralLiquidity(ctx sdk.Context, msg *types.MsgRemoveUnilateralLiquidity) (sdk.Coins, error) {
+	var targetTokenDenom string
+
+	poolId := types.GetPoolId(msg.CounterpartyDenom)
+	pool, exist := k.GetPool(ctx, poolId)
+	if exist != true {
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:386]: if  !exist

File: modules/coinswap/keeper/keeper.go [hunk 7]
```diff
@@ -282,6 +378,113 @@ func (k Keeper) removeLiquidity(ctx sdk.Context, poolAddr, sender sdk.AccAddress
...
+
+	if lptBalanceAmt.LT(msg.ExactLiquidity) {
+		return sdk.Coins{}, sdkerrors.Wrapf(types.ErrInsufficientFunds,
+			fmt.Sprintf("insufficient %s funds, user expected: %s, actual: %s",
+				lptDenom, msg.ExactLiquidity.String(), lptBalanceAmt.String()))
+	}
+
+	if lptBalanceAmt.Equal(msg.ExactLiquidity) {
+		return sdk.Coins{}, sdkerrors.Wrapf(types.ErrConstraintNotMet,
+			fmt.Sprintf("forbid to withdraw all liquidity unilaterally"))
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:424]: no args ?

File: modules/coinswap/keeper/keeper.go (hunk 1 above)
Comment (taramakage) [reply modules/coinswap/keeper/keeper.go:117]: Proposed to do that way before but was asked to just add an interface assertion. Maybe refactor this now?

File: modules/coinswap/keeper/keeper.go [hunk 8]
```diff
@@ -282,6 +378,113 @@ func (k Keeper) removeLiquidity(ctx sdk.Context, poolAddr, sender sdk.AccAddress
...
+	}
+
+	if targetBalanceAmt.LT(msg.MinToken.Amount) {
+		return sdk.Coins{}, sdkerrors.Wrapf(types.ErrInsufficientFunds,
+			fmt.Sprintf("insufficient %s funds, user expected: %s, actual: %s",
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper.go:429]: `fmt.Sprintf` is duplicate code, refer `sdkerrors.Wrapf`

File: modules/coinswap/keeper/keeper_test.go [hunk 9]
```diff
@@ -109,10 +109,10 @@ func setupWithGenesisAccounts(t *testing.T) *simapp.SimApp {
 }
 
 func (suite *TestSuite) TestLiquidity() {
-	btcAmt, _ := sdk.NewIntFromString("100")
-	standardAmt, _ := sdk.NewIntFromString("10000000000000000000")
+	btcAmt, _ := sdkmath.NewIntFromString("100")
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/keeper_test.go:112]: replace all `sdk.Int` with `sdkmath.Int`

File: modules/coinswap/keeper/migrations.go [hunk 10]
```diff
@@ -5,6 +5,7 @@ import (
 
 	v150 "github.com/irisnet/irismod/modules/coinswap/migrations/v150"
 	v152 "github.com/irisnet/irismod/modules/coinswap/migrations/v152"
+	v153 "github.com/irisnet/irismod/modules/coinswap/migrations/v153"
 )
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/migrations.go:9]: rename package name `v150`,`v152`.`v153` to `v1`,`v2`,'v3', consistent with moduleVersion

File: modules/coinswap/keeper/msg_server.go [hunk 11]
```diff
@@ -76,6 +101,35 @@ func (m msgServer) RemoveLiquidity(goCtx context.Context, msg *types.MsgRemoveLi
...
+	)
+
+	var coins = make([]*sdk.Coin, 0, withdrawCoins.Len())
+	for _, coin := range withdrawCoins {
+		coins = append(coins, &coin)
+	}
+
+	return &types.MsgRemoveUnilateralLiquidityResponse{
+		WithdrawCoins: coins,
+	}, nil
```
Comment (dreamer-zq) [inline modules/coinswap/keeper/msg_server.go:130]: no need array point,

File: modules/coinswap/simulation/operations.go [hunk 12]
```diff
@@ -14,16 +11,22 @@ import (
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	simtypes "github.com/cosmos/cosmos-sdk/types/simulation"
 	"github.com/cosmos/cosmos-sdk/x/simulation"
+	"math/big"
+	"math/rand"
+	"strings"
+	"time"
```
Comment (dreamer-zq) [inline modules/coinswap/simulation/operations.go:17]: format code

File: proto/coinswap/tx.proto [hunk 13]
```diff
@@ -71,6 +104,30 @@ message MsgRemoveLiquidityResponse {
...
+  ];
+  int64 deadline = 4;
+  string sender = 5;
+}
+
+// MsgRemoveUnilateralLiquidityResponse defines the Msg/RemoveLiquidity response
+// type
+message MsgRemoveUnilateralLiquidityResponse {
+  repeated cosmos.base.v1beta1.Coin withdraw_coins = 1;
+}
```
Comment (dreamer-zq) [inline proto/coinswap/tx.proto:129]: repeated cosmos.base.v1beta1.Coin withdraw_coins = 1  [(gogoproto.nullable) = false];

**Pull request merged**